### ISSUE TITLE
Add support for .clang_complete files.

### ISSUE THREAD

**Issue by promethe42** [joeroback/atom-clang#8]
Hello,

AFAIK .clang_complete files are a "standard" way to define clangrelated auto-completion options.
They can easily be generated by the build system so devs don't have to setup anything.
Other [Atom packages support it](https://github.com/AtomLinter/linter-clang#clang_complete) and they are [not very hard to support](https://github.com/promethe42/linter-clang/commit/69106d9e3c16c973f28c47d3d07a292abb3dc973).

Do you have any plan for supporting .clang_complete files?

Regards,

**Comment by joeroback** [comment]
oh i didn't know that was a thing. i was planning on support complie_commands.json but i am currently trying to figure out how to handle header-only files, as compile_commands.json only has the actual compilations, so when a header-only file (looks of std and boost) is opened, its unclear from libclang perspective how to handle this...

in any case, i will add supporting .clang_complete files as well. seems like an easier and more standard way to support project-level flags.

thanks!

**Comment by joeroback** [comment]
so as i plan this out more, my initial tests of doing things like directory searches in javascript/coffee in atom's UI thread, its slow.. too slow. 3x-4x slower than in C++.. so I am leaning towards pushing the .clang_complete search and file IO into the c++ side of things..

Test were done a few machines

- macOS 10.12.2, Late 2016 Macbook Pro 15-inch, i7-6920HQ
- Ubuntu 16.04, Dell Precision 5510, i7-6820HQ
- Ubuntu 14.04, i7-5930k (overclocked to 4.4GHz)

So those are high-end machines, I hate to see the numbers are slower machines. I 100% want to support this, as I would like to start using this myself, but I do not want to increase the code completion latency 2-3x just searching for .clang_complete files.

Only, I would still like to keep the possibility of supporting Windows soon. So any directory/file IO needs to be cross platform. Boost Filesystem could be that, but than boost would be in install requirement... I mean most people installing a C/C++ code completer probably use boost or at least are familiar with it... In only C++17 was here...

**Comment by promethe42** [comment]
> I do not want to increase the code completion latency 2-3x just searching for .clang_complete files

Can't you just keep a cache table file => .clang_complete path so that the search is done only once per-file ?

**Comment by joeroback** [comment]
I could do that, but then if one modifies the .clang_complete file or adds a new one, you would have to reload the project,   which I think is kinda lame..

**Comment by joeroback** [comment]
I continued doing more tests on Linux and macOS. Using C++ and boost is so much faster searching the directories, if I use C++ for it, on every parse attempt, I could search the project directories without any real noticeable increase in latency. Being as optimal as I could in coffee/javascript, it added 175ms - 250ms delay only searching 2-5 directories deep. Additionally doing it in C++ means it going to happen outside the Atom IO thread, on one of the libuv threads.

Since there are many factors that cause a full parse (clang flags needed) vs reparse (clang flags are not needed for reparse), and the logic of whether to have to fully parse vs reparse is somewhat deep in the c++. I can delay doing any flag searching (of any kind) until the c++ code is 100% sure it wants to do a full parse. During editing and code completion, it's nearly 100% reparse calls.

i also figured out how to build a node module with CMake using https://github.com/cmake-js/cmake-js so it will be easy to add boost as a dependency for the filesystem stuff and stay cross platform. https://github.com/cmake-js/boost-lib is also an option that if system boost isn't found, it actually downloads and compiles boost (only the components used, filesystem in this case).

So I am closing in on a final design that will support 4 types of clang flags, whose preference is in this order:
1. compile_commands.json
2. .clang_complete file
3.atom project config (from package project-manager)
4. atom global config

An example search will start looking for compile_commands.json database (one per project), then in the current source directory looking for a .clang_complete file and go out until it hits a project root directory. Finally looking in the Atom config. For example, looking for clang flags for `/home/john/projects/MyProject/resources/lib/src/source1.cpp`, with `/home/john/projects/MyProject` being the project root, the search would be

```
1. search compile_commands.json
2. /home/john/projects/MyProject/resources/lib/src/.clang_complete
3. /home/john/projects/MyProject/resources/lib/.clang_complete
4. /home/john/projects/MyProject/resources/.clang_complete
5. /home/john/projects/MyProject/.clang_complete
6. project config via project-manager
7. global atom.config
```

Of course all the searching and flags would be cached in different ways, as to only do the searching, reading, parsing of flags, when something changes.

**Comment by joeroback** [comment]
To give some examples of searching a project that is 15 directories deep and not finding a `.clang_complete` file (worst case).

```
$ /tmp/timecycles /tmp/boost_test
0.008	24211504
$ /tmp/timecycles coffee /tmp/coffee_test_sync.coffee
0.203	591056569
$ /tmp/timecycles coffee /tmp/coffee_test_async.coffee 
0.163	474817939
```
First column is seconds as reported by gettimeofday() and second column is cycles as reported by rdtscp() (x86 timestamp counter register)

There are two coffee implementations. Sync uses all synchronous calls from the 'fs' module. Async uses 100% asynchronous calls and promises. And this is just a test for finding a `.clang_complete` file. Reading and parsing the file in C++ has similar speedups, in addition, at some point the flags have to go from JavaScript to C++, which would require more processing to convert the data types.

Using C++ I think we could have a better user experience by being able to detect automatically when things change, faster parsing, etc.

**Comment by joeroback** [comment]
oh I am kinda embarrassed, I just realized two errors with this timing...

1. I am timing the overhead of running "coffee", which in the case of atom, the coffee/javascript engine will already be loaded... So I changed to using console.time/console.timeEnd to time the coffee versions.
2. The first column is not seconds, but milliseconds.

First column calculation in `/tmp/timecycles`
```
// calculate milliseconds
ms  = (tv2.tv_sec - tv1.tv_sec) * 1000;
ms += (tv2.tv_usec - tv1.tv_usec) / 1000;
```

So rechecking the timing, C++ is still the clear winner, but coffee sync is more than acceptable at < 1ms. Coffee async is probably a lot of overhead of async + promises. Averages over about 30 runs
```
boost        =  0.008 ms
coffee sync  =  0.573 ms
coffee async = 10.156 ms
```

**Comment by xphoenix** [comment]
That for the great plugin on the first place :)

In case of cmake compile_commands.json could be (and usually is) outside of the project root, so I would suggest to add setting "where to commands database is located" and use hierarchy search only if it is not set.

As for headers only files have no idea how could flags be possible calculated automatically. But I would be happy if atom-clang use following logic:
1. If compilcation_commands present and file is present there - use that flags
2. Otherwise use "default" flags: [.clang_complete, project_config, globals]

It would be really nice to get compile_commands.json work at least it solves most of problems, headers are important but somehow parallel to this functionality, IMHO
